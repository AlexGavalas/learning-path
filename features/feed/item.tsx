import { useState, FormEventHandler } from 'react';
import { format } from 'date-fns';

import { Dialog } from '@components/dialog';
import { Button } from '@components/button';
import { Textarea } from '@components/textarea';

interface ListItemProps {
    post: Post;
    isMine: boolean;
    handleDelete: (id: string) => void;
    onPostUpdate: (id: string, newPost: string) => Promise<void>;
}

export const ListItem = ({
    post,
    isMine,
    handleDelete,
    onPostUpdate,
}: ListItemProps) => {
    const [openEditDialog, setOpenEditDialog] = useState(false);
    const { post: postData, name, id, created_at } = post;

    const postDate = format(new Date(created_at), 'dd/MM/yyyy HH:mm');

    const closeDialog = () => setOpenEditDialog(false);

    const updatePost: FormEventHandler<HTMLFormElement> = async (e) => {
        e.preventDefault();

        const newPost = new FormData(e.currentTarget)
            .get('newpost')
            ?.toString();

        if (!newPost) return;

        await onPostUpdate(post.id, newPost);

        closeDialog();
    };
    
    return (
        <li className="list-none py-6 border-0 border-solid last:border-b-0 flex flex-col">
            <p className="border-solid border-0 border-l-2 border-l-teal-500 dark:border-l-yellow-500 pl-4 break-words">
                {postData}
            </p>
            <div className="self-end flex gap-4 items-center">
                {isMine && (
                    <div className="flex gap-2 h-5">
                        <Button
                            variant="danger"
                            onClick={() => handleDelete(id)}
                        >
                            Delete
                        </Button>
                        <Button onClick={() => setOpenEditDialog(true)}>
                            Edit
                        </Button>
                    </div>
                )}
                <p className="italic text-sm ">
                    - {name} at {postDate}
                </p>
            </div>
            <Dialog
                open={openEditDialog}
                size="md"
                onClickOutside={closeDialog}
            >
                <p className="m-0 mb-2">Make your changes</p>
                <form onSubmit={updatePost}>
                    <Textarea name="newpost" defaultValue={post.post} />
                    <div className="flex gap-2 justify-end">
                        <Button onClick={closeDialog} variant="danger">
                            Cancel
                        </Button>
                        <Button>Save</Button>
                    </div>
                </form>
            </Dialog>
        </li>
    );
};
