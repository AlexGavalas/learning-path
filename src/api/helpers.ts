import axios from 'axios';

export const fetchFileFromStorage = async (
    filename: string,
): Promise<string> => {
    const fileServerUrl = String(process.env.PUBLIC_FILE_SERVER_URL);

    const response = await axios.get(`${fileServerUrl}/${filename}`, {
        responseType: 'text',
    });

    return response.data as string;
};
