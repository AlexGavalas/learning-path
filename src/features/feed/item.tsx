import { format } from 'date-fns';
import { type FormEventHandler, useState } from 'react';

import { Button } from '~components/button';
import { Dialog } from '~components/dialog';
import { Textarea } from '~components/textarea';

import { type UserPost } from '../../../types/notes.types';

interface ListItemProps {
    post: UserPost;
    isMine: boolean;
    onPostDelete: (id: string) => void;
    onPostUpdate: (id: string, newPost: string) => Promise<void>;
}

export const ListItem = ({
    post,
    isMine,
    onPostDelete,
    onPostUpdate,
}: ListItemProps) => {
    const [openEditDialog, setOpenEditDialog] = useState(false);
    const { post: postData, name, id, created_at } = post;

    const postDate = format(new Date(created_at), 'dd/MM/yyyy HH:mm');

    const closeDialog = () => setOpenEditDialog(false);

    const updatePost: FormEventHandler<HTMLFormElement> = async (e) => {
        e.preventDefault();

        const newPost = new FormData(e.currentTarget)
            .get('newpost')
            ?.toString();

        if (!newPost) return;

        await onPostUpdate(post.id, newPost);

        closeDialog();
    };

    return (
        <li className="flex list-none flex-col border-0 border-solid py-6 last:border-b-0">
            <p className="break-words border-0 border-l-2 border-solid border-l-teal-500 pl-4 dark:border-l-yellow-500">
                {postData}
            </p>
            <div className="flex items-center gap-4 self-end">
                {isMine && (
                    <div className="flex h-5 gap-2">
                        <Button
                            variant="danger"
                            onClick={() => onPostDelete(id)}
                        >
                            Delete
                        </Button>
                        <Button
                            onClick={(e) => {
                                e.stopPropagation();
                                setOpenEditDialog(true);
                            }}
                        >
                            Edit
                        </Button>
                    </div>
                )}
                <p className="text-sm italic ">
                    - {name} at {postDate}
                </p>
            </div>
            <Dialog
                open={openEditDialog}
                size="md"
                onClickOutside={closeDialog}
            >
                <form onSubmit={updatePost} className="flex flex-col gap-2">
                    <Textarea
                        label="Make your changes"
                        name="newpost"
                        defaultValue={post.post}
                    />
                    <div className="flex justify-end gap-2">
                        <Button
                            onClick={closeDialog}
                            type="button"
                            variant="danger"
                        >
                            Cancel
                        </Button>
                        <Button>Save</Button>
                    </div>
                </form>
            </Dialog>
        </li>
    );
};
