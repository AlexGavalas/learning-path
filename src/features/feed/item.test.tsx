import { renderWithUser, screen } from '~test/helpers';

import { ListItem } from './item';

const post: UserPost = {
    id: '1',
    created_at: new Date('2022-01-01').toISOString(),
    name: 'name',
    post: 'post',
};

const renderListItem = ({ isMine }: { isMine: boolean }) => {
    const mockOnPostDelete = jest.fn();
    const mockOnPostUpdate = jest.fn();

    const renderResult = renderWithUser(
        <ListItem
            post={post}
            isMine={isMine}
            onPostDelete={mockOnPostDelete}
            onPostUpdate={mockOnPostUpdate}
        />,
    );

    return {
        ...renderResult,
        mockOnPostDelete,
        mockOnPostUpdate,
    };
};

describe('<ListItem />', () => {
    describe('when the post is not mine', () => {
        it('renders', () => {
            const { container } = renderListItem({ isMine: false });

            expect(container).toMatchSnapshot();
        });

        it('does not expose controls', () => {
            renderListItem({ isMine: false });

            expect(screen.queryByRole('button')).not.toBeInTheDocument();
        });
    });

    describe('when the post is mine', () => {
        it('renders', () => {
            const { container } = renderListItem({ isMine: true });

            expect(container).toMatchSnapshot();
        });

        it('exposes two buttons', () => {
            renderListItem({ isMine: true });

            expect(screen.queryAllByRole('button')).toHaveLength(2);
        });

        describe('when user clicks delete', () => {
            it('calls onPostDelete', async () => {
                const { user, mockOnPostDelete } = renderListItem({
                    isMine: true,
                });

                await user.click(
                    screen.getByRole('button', { name: /delete/i }),
                );

                expect(mockOnPostDelete).toHaveBeenCalledTimes(1);
                expect(mockOnPostDelete).toHaveBeenCalledWith(post.id);
            });
        });

        describe('when user clicks edit', () => {
            it('opens edit dialog', async () => {
                const { user } = renderListItem({
                    isMine: true,
                });

                await user.click(screen.getByRole('button', { name: /edit/i }));

                expect(screen.getByRole('dialog')).toBeInTheDocument();
            });

            describe('when user types and clicks save', () => {
                it('calls onPostUpdate', async () => {
                    const { user, mockOnPostUpdate } = renderListItem({
                        isMine: true,
                    });

                    await user.click(
                        screen.getByRole('button', { name: /edit/i }),
                    );

                    const textInput = screen.getByRole('textbox');
                    const addedText = 'abc';
                    const finalPost = `${post.post}${addedText}`;

                    await user.type(textInput, addedText);

                    expect(textInput).toHaveValue(finalPost);

                    await user.click(
                        screen.getByRole('button', { name: /save/i }),
                    );

                    expect(mockOnPostUpdate).toHaveBeenCalledTimes(1);
                    expect(mockOnPostUpdate).toHaveBeenCalledWith(
                        post.id,
                        finalPost,
                    );
                });
            });

            describe('when user clears the input and clicks save', () => {
                it('does not call onPostUpdate', async () => {
                    const { user, mockOnPostUpdate } = renderListItem({
                        isMine: true,
                    });

                    await user.click(
                        screen.getByRole('button', { name: /edit/i }),
                    );

                    const textInput = screen.getByRole('textbox');

                    await user.clear(textInput);

                    await user.click(
                        screen.getByRole('button', { name: /save/i }),
                    );

                    expect(mockOnPostUpdate).not.toHaveBeenCalled();
                });
            });
        });
    });
});
