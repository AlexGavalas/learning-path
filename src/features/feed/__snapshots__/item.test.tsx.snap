// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ListItem /> when the post is mine renders 1`] = `
<div>
  <li
    class="flex list-none flex-col border-0 border-solid py-6 last:border-b-0"
  >
    <p
      class="break-words border-0 border-l-2 border-solid border-l-light-primary pl-4 dark:border-l-dark-primary"
    >
      post
    </p>
    <div
      class="flex items-center gap-4 self-end"
    >
      <div
        class="flex h-5 gap-2"
      >
        <button
          class="items-center rounded border-0 text-white p-2 flex bg-red-500 hover:bg-red-700 dark:bg-red-500 dark:hover:bg-red-700"
        >
          Delete
        </button>
        <button
          class="items-center rounded border-0 text-black dark:text-white p-2 flex bg-gray-200 hover:bg-gray-300 dark:bg-gray-500 dark:hover:bg-gray-600"
        >
          Edit
        </button>
      </div>
      <p
        class="text-sm italic "
      >
        - 
        name
         at 
        01/01/2022 02:00
      </p>
    </div>
  </li>
</div>
`;

exports[`<ListItem /> when the post is not mine renders 1`] = `
<div>
  <li
    class="flex list-none flex-col border-0 border-solid py-6 last:border-b-0"
  >
    <p
      class="break-words border-0 border-l-2 border-solid border-l-light-primary pl-4 dark:border-l-dark-primary"
    >
      post
    </p>
    <div
      class="flex items-center gap-4 self-end"
    >
      <p
        class="text-sm italic "
      >
        - 
        name
         at 
        01/01/2022 02:00
      </p>
    </div>
  </li>
</div>
`;
